Tree-sync friendly lookup sync tests

dapplion opened the pull request:
## Issue Addressed

- Step 0 of the tree-sync roadmap https://github.com/sigp/lighthouse/issues/7678

Current lookup sync tests are written in an explicit way that assume how the internals of lookup sync work. For example the test would do:

- Emit unknown block parent message
- Expect block request for X
- Respond with successful block request
- Expect block processing request for X
- Response with successful processing request
- etc..

This is unnecessarily verbose. And it will requires a complete re-write when something changes in the internals of lookup sync (has happened a few times, mostly for deneb and fulu).

What we really want to assert is:

- WHEN: we receive an unknown block parent message
- THEN: Lookup sync can sync that block
- ASSERT: Without penalizing peers, without unnecessary retries 

## Proposed Changes

Keep all existing tests and add new cases but written in the new style described above. The logic to serve and respond to request is in this function `fn simulate` https://github.com/dapplion/lighthouse/blob/2288a3aeb11164bb1960dc803f41696c984c69ff/beacon_node/network/src/sync/tests/lookups.rs#L301 
- It controls peer behavior based on a `CompleteStrategy` where you can set for example "respond to BlocksByRoot requests with empty"
- It actually runs beacon processor messages running their clousures. Now sync tests actually import blocks, increasing the test coverage to the interaction of sync and the da_checker.
- To achieve the above the tests create real blocks with the test harness. To make the tests as fast as before, I disabled crypto with `TestConfig`

Along the way I found a couple bugs, which I documented on the diff.

## Review guide

Look at `lighthouse/beacon_node/network/src/sync/tests/lookups.rs` directly (no diff). 

Other changes are very minor and should not affect production paths

dapplion committed:
WIP Add more sync tests

dapplion committed:
Add more sync tests

dapplion committed:
Add blobs and columns

dapplion committed:
Disable crypto programatically

dapplion committed:
Fix more tests

dapplion committed:
Fix bug with mock execution engine

dapplion committed:
Fix tests

dapplion committed:
Fix remaining tests

dapplion committed:
Add more fulu tests

dapplion committed:
Merge remote-tracking branch 'sigp/unstable' into more-sync-tests

dapplion committed:
Lint full

dapplion committed:
Enable ignored test

dapplion committed:
Allow to reorder event order

dapplion committed:
Tests passing

dapplion committed:
Lint

dapplion committed:
Run phase0 only for network tests

dapplion commented on beacon_node/network/src/sync/block_lookups/mod.rs:
```diff
@@ -743,6 +766,15 @@ impl<T: BeaconChainTypes> BlockLookups<T> {
+            // removed from the da_checker. Note that ALL components are removed from the da_checker
+            // so when we re-download and process the block we get the error
+            // MissingComponentsAfterAllProcessed and get stuck.
+            lookup.reset_requests();
```
Bug found on testing, lookups may be stuck given this sequence of events

dapplion commented on beacon_node/network/src/sync/manager.rs:
```diff
@@ -512,17 +514,18 @@ impl<T: BeaconChainTypes> SyncManager<T> {
+        // sending retry requests to the disconnecting peer.
+        for sync_request_id in self.network.peer_disconnected(peer_id) {
+            self.inject_error(*peer_id, sync_request_id, RPCError::Disconnected);
+        }
```
Minor bug, we need to remove the peer from the sync states (e.g. `self.block_lookups`) then inject the disconnect events. Otherwise we may send requests to peers that are already disconnected. I don't think there's risk of sync getting stuck if libp2p rejects sending messages to disconnected peers, but deserves a fix anyway.

dapplion committed:
Reduce diff

dapplion committed:
Reduce diff

dapplion committed:
Merge remote-tracking branch 'sigp/unstable' into more-sync-tests

dapplion committed:
Use FIFO

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -1158,1561 +1537,697 @@ fn stable_rng() {
 
 // Regression test for https://github.com/sigp/lighthouse/pull/7118
 // 8042 UPDATE: block was previously added to the failed_chains cache, now it's inserted into the
 // ignored chains cache. The regression test still applies as the chaild lookup is not created
```
minor typo 

```suggestion
// ignored chains cache. The regression test still applies as the child lookup is not created
```

dapplion replied:
Done

jimmygchen commented on beacon_node/beacon_chain/src/chain_config.rs:
```diff
@@ -162,6 +172,10 @@ impl Default for ChainConfig {
             node_custody_type: NodeCustodyType::Fullnode,
+            test_config: TestConfig {
+                disable_crypto: false,
+                disable_fetch_blobs: false,
```
Does this currently do anything in the tests, other than skipping the attempt? 
The attempt will always fail because the mock-el does not support `getBlobs` right? 
I made an attempt to add it in #7986, but we didn't end up merging it because it didn't feel useful.

dapplion replied:
Removed `disable_fetch_blobs` as it does not materially increase speed (I think I added it to make the logs less messy) 


Also applied `disable_crypto` to block production in the TestRig

jimmygchen commented on beacon_node/beacon_chain/src/test_utils.rs:
```diff
@@ -783,7 +783,13 @@ where
+            .get_blinded_block(block_root)
+            .unwrap()
+            .unwrap_or_else(|| {
+                panic!("block root does not exist in external harness {block_root:?}")
```
This isn't always "external" harness

```suggestion
                panic!("block root does not exist in harness {block_root:?}")
```

dapplion replied:
Done

jimmygchen commented on beacon_node/network/src/sync/block_lookups/mod.rs:
```diff
@@ -117,15 +117,23 @@ pub struct BlockLookups<T: BeaconChainTypes> {
 
 #[cfg(test)]
+#[derive(Debug)]
 /// Tuple of `SingleLookupId`, requested block root, awaiting parent block root (if any),
```
Please update doc - no longer a tuple and awaiting parent block removed.

dapplion replied:
Done

jimmygchen commented on beacon_node/network/Cargo.toml:
```diff
@@ -58,6 +58,7 @@ gossipsub = { workspace = true }
 k256 = "0.13.4"
 kzg = { workspace = true }
 matches = "0.1.8"
+paste = "1.0.15"
```
```suggestion
paste = { workspace = true }
```

jimmygchen commented on Makefile:
```diff
@@ -36,8 +36,12 @@ PROFILE ?= release
 RECENT_FORKS_BEFORE_GLOAS=electra fulu
 
 # List of all recent hard forks. This list is used to set env variables for http_api tests
+# Include phase0 to test the code paths in sync that are pre blobs
```
We already have `nightly-tests` that runs prior fork tests
https://github.com/sigp/lighthouse/pull/8319

But i just realised it hasn't been activated on the `sigp` fork because github only run scheduled workflows from the main branch (`stable`), we can either wait until the release or have a separate PR to `stable` to activate this.

jimmygchen replied:
Made a PR to activate these nightly tests:
https://github.com/sigp/lighthouse/pull/8636

dapplion replied:
Could we keep this for network tests only? It's just one extra fork and makes it easy to debug and catch errors. For sync tests we should keep the forks that add new objects like run only 
- phase0, deneb, fulu

jimmygchen commented on Makefile:
```diff
@@ -226,10 +230,10 @@ test-op-pool-%:
 
 test-network-%:
-	env FORK_NAME=$* cargo nextest run --release \
+	env FORK_NAME=$* cargo nextest run --no-fail-fast --release \
```
Do you want to merge this, or just for your local testing? I think it's fine to not fail fast, as long as the job doens't take forever to run, e.g. beacon-chain-tests

dapplion replied:
The --no-fail-fast gives you more information on CI on which set of tests failed. A single fork run is not that long so we don't save that much time. But the full report is useful

jimmygchen replied:
yeah thats reasonable, thanks

jimmygchen commented on beacon_node/network/src/sync/tests/mod.rs:
```diff
@@ -58,17 +65,102 @@ struct TestRig {
     /// Beacon chain harness
     harness: BeaconChainHarness<EphemeralHarnessType<E>>,
+    /// External beacon chain harness to produce blocks that are not imported
+    external_harness: BeaconChainHarness<EphemeralHarnessType<E>>,
```
Any reason to have this as a field on `TestRig`? I see that it's only used in `build_chain`

dapplion replied:
Good find! Moved to build_chain

jimmygchen commented on beacon_node/network/src/sync/manager.rs:
```diff
@@ -512,17 +514,18 @@ impl<T: BeaconChainTypes> SyncManager<T> {
 
+        // Inject a Disconnected error on all requests associated with the disconnected peer
+        // to retry all batches/lookups. Only after removing the peer from the data structures to
+        // sending retry requests to the disconnecting peer.
```
Missing word i think 

```suggestion
        // avoid sending retry requests to the disconnecting peer.
```

dapplion committed:
Apply PR review

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -128,583 +242,1052 @@ impl TestRig {
-            None
+    // Network / external peers simulated behaviour
+
+    async fn simulate(&mut self, complete_strategy: CompleteStrategy) {
```
Would be good to add some brief docs here on what it does and how to use this.

The names `simulate` and `CompleteStrategy` are a bit generic. I thought about `NetworkResponseStrategy` or `PeerBehaviourStrategy` but realised the struct also has non peer behaviour related configuration, like `process_result_conditional` and `block_imported_while_processing`.  I can't think of a better name right now though.

dapplion replied:
Maybe `SimulateConfig`?

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -1,73 +1,177 @@
+        self
+    }
+
+    fn process_result<F>(mut self, f: F) -> Self
```
I think we can improve this name to make it obvious this is implementing builder pattern, rather than actually processing results, could use the standard `with_` prefix.

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -1,73 +1,177 @@
+        self
+    }
+
+    fn block_imported_while_processing(mut self, block_root: Hash256) -> Self {
```
same as above

jimmygchen commented on beacon_node/beacon_chain/src/test_utils.rs:
```diff
@@ -1410,9 +1418,12 @@ where
-                                &self.validator_keypairs[*validator_index].sk.sign(message),
-                            );
+                            // If disable_crypto is true keep the attestation signature as infinity
+                            if self.chain.config.test_config.disable_crypto {
```
I think we're missing a `!` here?

jimmygchen commented on beacon_node/beacon_chain/src/test_utils.rs:
```diff
@@ -1510,9 +1521,12 @@ where
-                                &self.validator_keypairs[*validator_index].sk.sign(message),
-                            );
+                            // If disable_crypto is true keep the attestation signature as infinity
+                            if self.chain.config.test_config.disable_crypto {
```
missing a `!` here?

jimmygchen commented on beacon_node/network/src/sync/tests/mod.rs:
```diff
@@ -68,7 +77,88 @@ struct TestRig {
+    /// Lookup's Id
+    id: Id,
+    block_root: Hash256,
+    max_seen_peers: HashSet<PeerId>,
```
i think `max_` isn't necessary here, or does it mean something other than `seen peers`?

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -1158,1561 +1532,697 @@ fn stable_rng() {
-    // Trigger the request
-    rig.trigger_unknown_block_from_attestation(block_hash, peer_id);
-    let id = rig.expect_block_lookup_request(block_hash);
+/// Assert that sync completes from a GossipUnknownParentBlob / UknownDataColumnParent
```
typo 

```suggestion
/// Assert that sync completes from a GossipUnknownParentBlob / UnknownDataColumnParent
```

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -1158,1561 +1532,697 @@ fn stable_rng() {
+/// Test added in https://github.com/sigp/lighthouse/commit/84c7d8cc7006a6f1f1bb5729ab222b9f85f72727
+/// TODO: This test was added on a very old version of lookup sync. It's unclear if the situation
+/// it wants to recreate is possible or problematic in current code. Skipping.
+#[ignore]
```
Is this test failing?

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -128,583 +242,1052 @@ impl TestRig {
-            chain_hash,
-            BlockProcessingResult::Ok(AvailabilityProcessingStatus::Imported(chain_hash)),
-        );
+    fn reset_metrics(&mut self) {
```
`reset_metrics` seems a bit misleading here, perhaps something like `capture_metrics_baseline` is more suitable?

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -128,583 +242,1052 @@ impl TestRig {
-        let id = self.find_single_lookup_for(self.find_oldest_parent_lookup(chain_hash));
-        self.single_blob_component_processed(id, result);
+    fn completed_lookups(&self) -> usize {
+        // Substract initial value to allow resetting metrics mid test
```
typo 

```suggestion
        // Subtract initial value to allow resetting metrics mid test
```

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -128,583 +242,1052 @@ impl TestRig {
+                self.sync_rx_queue.push(ev);
+            }
+
+            // Choose at random which queue to process first
```
I don't fully understand the purpose of deterministically choosing a random queue to process?

jimmygchen reviewed (changes requested):
Hey @dapplion 

I've done a round of review and have added a few comments. There's a logic error in the condition and the CI failures that needs to be fixed.

Overall I think the approach is good - offloading complexity to TestRig makes individual tests cleaner. I'm slightly concerned about the growing complexity of `TestRig` itself and how easy it would be to extend it in the future, but I guess this may be needed as sync itself is inheritlenty complex.

dapplion committed:
Address PR review comments

- Fix logic error: add missing `!` in attestation signing conditions
- Fix typos: Substract -> Subtract, UknownDataColumnParent -> UnknownDataColumnParent
- Rename reset_metrics to capture_metrics_baseline
- Rename max_seen_peers to seen_peers
- Use with_ prefix for builder methods (with_process_result, with_block_imported_while_processing)
- Add documentation to simulate fn and CompleteStrategy

dapplion committed:
Use features to disable crypto

dapplion committed:
Reduce diff

dapplion committed:
Reduce diff

dapplion committed:
Fix fake crypto usage

dapplion committed:
Add crypto on test

dapplion committed:
Fix crypto

dapplion committed:
fmt

dapplion committed:
Lint crypto_on

dapplion committed:
Complete crypto_on tests

dapplion committed:
Merge branch 'unstable' into more-sync-tests

dapplion committed:
lint

dapplion committed:
Lint with fake_crypto feature

dapplion committed:
Merge remote-tracking branch 'sigp/unstable' into more-sync-tests

dapplion committed:
rename CompleteStrategy to SimulateConfig

dapplion committed:
Merge remote-tracking branch 'sigp/unstable' into more-sync-tests

# Conflicts:
#	beacon_node/network/src/sync/tests/lookups.rs
#	beacon_node/network/src/sync/tests/mod.rs
#	beacon_node/network/src/sync/tests/range.rs

dapplion committed:
adapt tests to new RpcBlock enum API after merge

Update lookups.rs and range.rs to work with the new RpcBlock enum
(FullyAvailable | BlockOnly) from PR #8424. Replace removed methods
(.blobs(), .custody_columns(), new_without_blobs, new_with_custody_columns)
with the new block_data() + AvailableBlockData API. Also adapt to
superstruct-generated getters on DataColumnSidecar (Fulu/Gloas variants).

dapplion committed:
Merge remote-tracking branch 'sigp/unstable' into more-sync-tests

dapplion committed:
remove random queue processing from simulate()

dapplion committed:
remove stale test_same_chain_race_condition

dapplion committed:
remove unused import

dapplion committed:
remove unused block_imported_while_processing

dapplion committed:
add test for same chain race condition

Reproduces the race condition from PR #3677 where block 1 gets imported
via gossip right before the lookup processes it. The test verifies that
when the oldest block in a parent lookup chain returns DuplicateFullyImported,
the remaining blocks (2 and 3) still process successfully.

dapplion committed:
fix formatting

dapplion committed:
address review feedback: rename builder method and fix test docs

dapplion commented:
@jimmygchen Thanks for the review! I've addressed all your feedback:

**Already fixed:**
- Documentation, naming improvements (`max_seen_peers` → `seen_peers`, `with_process_result`)
- Removed unused code (`disable_fetch_blobs`, `external_harness` field, random queue selection)
- Fixed panic message and typos

**Just fixed (8d29e73a9):**
- Renamed `import_block_before_process` → `with_import_block_before_process` for builder pattern consistency  
- Fixed incomplete test documentation

All 12 review threads resolved. The 2 remaining threads are my own bug notes.

dapplion committed:
Merge branch 'unstable' into more-sync-tests

dapplion committed:
Merge remote-tracking branch 'sigp/unstable' into more-sync-tests

jimmygchen commented on beacon_node/genesis/src/interop.rs:
```diff
@@ -124,7 +124,7 @@ impl<E: EthSpec> InteropGenesisBuilder<E> {
                 };
 
-                data.signature = data.create_signature(&keypair.sk, spec);
+                data.signature = SignatureBytes::empty();
```
Double check if we need these changes in this file.
Should we feature gate this?

jimmygchen replied:
Potentialy delete this function in a future PR?

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -132,582 +236,1161 @@ impl TestRig {
+            self.active_single_lookups(),
         );
+        self.expect_empty_network();
+        self.expect_no_active_lookups();
```
Add documentation for naming conventions, e.g. 

* `.expect_` for mocking / setting up component behaviour
* `.assert_` for making assertion on behaviour

jimmygchen commented on Makefile:
```diff
@@ -36,8 +36,12 @@ PROFILE ?= release
+# Include phase0 to test the code paths in sync that are pre blobs
 RECENT_FORKS=electra fulu gloas
 
+# For network tests include phase0 to cover genesis syncing (blocks without blobs or columns)
```
is this necessary or can we just cover this in the nightly tests?

jimmygchen replied:
ah right, you responded above. No objection to this.

jimmygchen commented on consensus/state_processing/src/per_block_processing/process_operations.rs:
```diff
@@ -641,7 +649,7 @@ pub fn apply_deposit<E: EthSpec>(
         // The signature should be checked for new validators. Return early for a bad
         // signature.
-        if is_valid_deposit_signature(&deposit_data, spec).is_err() {
+        if verify_signatures.is_true() && is_valid_deposit_signature(&deposit_data, spec).is_err() {
```
do we need to add the same check to all other places where `is_valid_deposit_signature` is called? or would it be cleaner to use feature so we don't have to pass `VerifySignatures` around and remember to add this check whereever we verify signature?
* https://github.com/sigp/lighthouse/blob/765dcae0f3522621a2b3d81f61209ffcc852230f/consensus/state_processing/src/per_block_processing/process_operations.rs#L889
* https://github.com/sigp/lighthouse/blob/605ef8e8e6cbb9bc974b767bb8a35a445d5a1936/consensus/state_processing/src/per_epoch_processing/single_pass.rs#L396

dapplion committed:
address pr review: rename expect_ to assert_, revert VerifySignatures threading, cfg(test) metrics

dapplion commented:
The `network-tests` CI failure is a pre-existing flaky test unrelated to this PR's changes:

```
data_column_reconstruction_at_deadline
  left: ["column_reconstruction", "gossip_data_column_sidecar", "worker_freed"]
 right: ["gossip_data_column_sidecar", "worker_freed", "nothing_to_do"]
```

The test has a latent race: on the last loop iteration, enqueuing the final gossip column triggers reconstruction nearly simultaneously. The `assert_event_journal_completes` in the loop expects to drain `gossip_data_column_sidecar` first, but `column_reconstruction` can fire before it. All merge queue runs on unstable pass — the race just surfaces under different CI timing/load. Worth hardening the event ordering assertion in a follow-up.

jimmygchen commented:
> The `network-tests` CI failure is a pre-existing flaky test unrelated to this PR's changes:
> 
> ```
> data_column_reconstruction_at_deadline
>   left: ["column_reconstruction", "gossip_data_column_sidecar", "worker_freed"]
>  right: ["gossip_data_column_sidecar", "worker_freed", "nothing_to_do"]
> ```
> 
> The test has a latent race: on the last loop iteration, enqueuing the final gossip column triggers reconstruction nearly simultaneously. The `assert_event_journal_completes` in the loop expects to drain `gossip_data_column_sidecar` first, but `column_reconstruction` can fire before it. All merge queue runs on unstable pass — the race just surfaces under different CI timing/load. Worth hardening the event ordering assertion in a follow-up.

are you @dapplion

dapplion committed:
fix flaky reconstruction_at_deadline test

jimmygchen commented on beacon_node/network/src/sync/tests/lookups.rs:
```diff
@@ -1161,1558 +1600,803 @@ fn stable_rng() {
+    r.corrupt_last_block_signature();
+    r.trigger_with_last_block();
+    r.simulate(SimulateConfig::happy_path()).await;
+    if cfg!(feature = "fake_crypto") {
```
Having this condition in the same test feels a bit confusing, as we're testing different outcomes and in the case of `fake_crypto` scenario, it doesn't match what the test name suggests,

I prefer something like:
```rust
  #[tokio::test]
  #[cfg(feature = "fake_crypto")]
  async fn block_signature_not_checked_with_fake_crypto() { ... }

  #[tokio::test]
  #[cfg(not(feature = "fake_crypto"))]
  async fn invalid_block_signature_fails_processing() { ... }
```

but yeah this is a minor style issue

jimmygchen reviewed (approved):
Hey @dapplion 

Nice work! This is a great improvement to the existing tests, which is becoming hard to maintain and will break once we change how sync works. 

It looks pretty solid now and we can keep iterating over the test interfaces as we add more tests.
